invenio_app_ils/permissions.py

```diff
@@ -89,6 +97,73 @@ def __init__(self, record):
         )
 
 
+class LoanExtendPermission(Permission):
+    """Return Permission to evaluate if the user can extend the loan."""
+
+    def __init__(self, record):
+        """Constructor."""
+        self.record = record
+
+        # NOTE: extra validation if the user is not librarian or admin
+        if current_user.id == int(record["patron_pid"]):
+            self.validate()
+
+        super().__init__(
+            UserNeed(int(record["patron_pid"])), backoffice_access_action
+        )
+
+    def validate(self):
+        """Validate if the loan can be extended."""
+        if not self.validate_extend_enabled():
+            raise InvalidLoanExtendError(
+                "You can extend this loan {} days before it "
+                "expires!".format(
+                    current_app.config["ILS_LOAN_WILL_EXPIRE_DAYS"])
+            )
+
+        if not self.validate_max_extensions():
+            raise InvalidLoanExtendError(
+                "You have reached the max number of extensions for THIS loan!"
+            )
+
+        if not self.validate_pending_loans():
+            raise InvalidLoanExtendError(
+                "There is a high demand for this literature!"
+            )
+
+        if self.validate_is_overdue():
+            raise InvalidLoanExtendError(
+                "This loan is overdue, its end date has passed!"
+            )
+
```

Comment:
I enumerated all the checks hopefully in a clear way in the issue, see: https://github.com/inveniosoftware/invenio-app-ils/issues/520

For the error messages, well they should not appear in the frontsite because there are checks there, but they appear in the APIs.

More I think about it and more I convince myself that we should have implemented errors in a different way:
* API consumers, e.g. a script or another app, should get a more technical message with a code
* frontsite, basically user interface, should probably ignore the technical message and use the code to re-map the error to something human readable and undestandable. WDYT?

Reply:
I agree with that. We should not compromise the technical description for just more user friendly errors in the frontend :)